mirage-router/unikernel.ml

```diff
@@ -0,0 +1,271 @@
+                         e Cstruct.hexdump_pp buf);
+            Lwt.return_unit
+        end
+      | Error e ->
+        Logs.err (fun m -> m "couldn't decode ethernet packet %s: %a"
+                     e Cstruct.hexdump_pp buf);
+        Lwt.return_unit
+      | Ok _ -> our_listen buf
+    in
+    N.listen net ~header_size:Ethernet_wire.sizeof_ethernet listen >|= function
+    | Error e ->
+      Logs.warn (fun m -> m "error %a listening on private network" N.pp_error e)
+    | Ok () ->
+      Logs.info (fun m -> m "listening on private network finished")
+
+  (* packets received over the tunnel *)
+  let rec ovpn_recv t =
+    O.read t.ovpn >>= fun datas ->
+    let ts = M.elapsed_ns () in
+    Lwt_list.fold_left_s (fun c data ->
+        match Ipv4_packet.Unmarshal.of_cstruct data with
+        | Ok (hdr, payload) ->
+          let c, pkt = Fragments.process c ts hdr payload in
+          begin match pkt with
+            | None -> ()
+            | Some (hdr, payload) ->
+              if local_network hdr.Ipv4_packet.dst then
+                match Ipv4_packet.(Unmarshal.int_to_protocol hdr.proto) with
+                | None ->
+                  Logs.warn (fun m -> m "ignoring %a (cannot decode IP protocol number)"
+                                Ipv4_packet.pp hdr)
+                | Some proto ->
+                  match forward_or_reject hdr payload (I.mtu t.private_ip) with
+                  | Ok hdr ->
+                    Lwt.async (fun () ->
+                        (* The IPv4.write here takes care of fragmenting the packet*)
+                        I.write t.private_ip ~src:hdr.Ipv4_packet.src hdr.Ipv4_packet.dst
+                          proto (fun _ -> 0) [ payload ] >|= function
+                        | Ok () -> ()
+                        | Error e ->
+                          (* could send back host unreachable if this was an arp timeout *)
+                          Logs.err (fun m -> m "error %a while forwarding %a"
+                                       I.pp_error e Ipv4_packet.pp hdr))
+                  | Error (`Icmp pay) ->
+                    (* send icmp error back via ovpn *)
+                    let hdr = {
+                      hdr with Ipv4_packet.ttl = 255 ;
+                               src = List.hd (I.get_ip t.private_ip) ; (* or which ip should be used? *)
+                               dst = hdr.Ipv4_packet.src ;
+                               proto = Ipv4_packet.Marshal.protocol_to_int `ICMP
+                    } in
+                    let payload_len = Cstruct.len pay in
+                    let hdr_cs = Ipv4_packet.Marshal.make_cstruct ~payload_len hdr in
+                    Lwt.async (fun () ->
+                      O.write t.ovpn (Cstruct.append hdr_cs pay) >|= fun _ ->
+                      ())
+                  | Error `Drop -> ()
+              else
+                Logs.warn (fun m -> m "ignoring %a (not for our network)"
```

Comment:
maybe useful to also pp `local_network` here?

Reply:
done in 742d14a